core/src/main/scala/org/apache/spark/sql/execution/joins/LocalJoin.scala

```diff
@@ -112,22 +116,41 @@ case class LocalJoin(leftKeys: Seq[Expression],
 
     // materialize dependencies in the entire buildRDD graph for
     // buildRDD.iterator to work in the compute of mapPartitionsPreserve below
-    materializeDependencies(buildRDD, new mutable.HashSet[RDD[_]]())
-
-    val schema = buildPlan.schema
-    streamRDD.mapPartitionsPreserveWithPartition { (context, _, itr) =>
-      val start = System.nanoTime()
-      val hashed = HashedRelationCache.get(schema, buildKeys, buildRDD,
-        buildPartition, context)
-      buildTime += (System.nanoTime() - start) / 1000000L
-      val estimatedSize = hashed.estimatedSize
-      buildDataSize += estimatedSize
-      context.taskMetrics().incPeakExecutionMemory(estimatedSize)
-      context.addTaskCompletionListener(_ => hashed.close())
-      join(itr, hashed, numOutputRows)
+    if (buildRDD.partitions.size == 1) {
+      materializeDependencies(buildRDD, new mutable.HashSet[RDD[_]]())
```

Comment:
Yes, you are correct. I tested materializeDependencies it does not work for the same reason mentioned by you.